rfcs/2022-02-11-responsive-and-adaptive-layout-strategy.md

```diff
@@ -0,0 +1,414 @@
+
+  - For eg: all the props(ones that are affected by responsive) for a component will then accept either single value or array/object of values for different screen sizes which is confusing and might lead to a lot of confusion at the API surface.
+
+    ```jsx
+    <Text fontSize="100" />
+
+    <Text fontSize={{ small: 100, medium: 100, large: 100, xl: 400 }} />
+    ```
+
+- Hence the idea here is that all these components except the Layout components will handle responsiveness based on predefined breakpoints internally without the consumers explicitly defining anything for them.
+
+### 2. Components that can give some flexibility to handle Responsiveness to consumers
+
+- While most of the components can handle responsiveness internally there are use cases for components where we would need to give flexibility to accept responsive values for certain props. For eg: Layout Components that handles the overall layout of the page might change the layout on different screen sizes based on the available space.
+  <img alt="Adaptive Layout with flex direction" src="./images/responsive-adaptive-layout/adaptive-layout-with-flex-direction.png">
+- The idea here is to have certain components that deal with Layout of the page expose some flexibility to consumers to tweak if they want the layout to look different on different screen sizes
+
+### 3. Components for Adaptive layout(rendering components conditionally)
+
+- After handling responsiveness internally and also giving flexibility for few scenarios there are still few complex cases that we need to handle:
+
+  - There are cases where we might need to render a completely different variation of a component based on the screen size and user experience. For eg: A Search Field component might be different for larger screen sizes and smaller screen sizes.
+    <img alt="Responsive Modal" src="./images/responsive-adaptive-layout/adaptive-search-field.png">
+
+  - Another example is to show and hide certain elements based on the screen size so that we can optimise for user experience for each screen size
+    <img alt="Responsive Modal" src="./images/responsive-adaptive-layout/adaptive-layout-with-multiple-layouts.png">
+
+- The idea here is to make it feasible for consumers to handle these kind of complex use cases with ease, maybe with a set of predefined components and APIs.
+
+## How will we actually implement the approaches discussed above?
+
+### On Blade Code Side
+
+- To begin with, as a first step we will first need to setup different screen sizes that we'll support and store it in tokens. This will handle our layout changes
+
+  ```
+  s:  480px // Mobile devices
+  m:  768px // Large mobile devices
+  l: 1024px // Small screens, Tablets
+  xl: 1200px // Desktops
+  ```
+
+- We would then need to add a listener in the `BladeProvider` which will listen to screen size changes and trigger a layout change.
```

Comment:
question: how will adaptive layout work with SSR? We would want to send only mobile bundle for mobile users but media query info is not available on server.